File: atcoder.jp/Python (PyPy 3.10-v7.3.12)/abc315/abc315_c.py
from collections import defaultdict

FS = defaultdict(list)
N = int(input())
for _ in range(N):
    F, S = map(int, input().split())
    FS[F].append(S)
for f in FS.keys():
    FS[f].sort(reverse=True)

ma = []
for s in FS.values():
    ma.append(s[0])
ma.sort(reverse=True)

ans = -1
if len(ma) >= 2:
    ans = ma[0] + ma[1]

for s in FS.values():
    if len(s) >= 2:
        ans = max(ans, s[0] + s[1] // 2)

print(ans)

File: atcoder.jp/Rust (rustc 1.70.0)/abc315/abc315_c.rs
use proconio::{fastout, input};

#[fastout]
fn main() {
    input! {
        n: usize,
        mut fs: [(usize, usize); n]
    }

    fs.sort_by_key(|(_, s)| *s);
    fs.reverse();

    let s = fs[0].1;
    let mut ans = 0;
    for i in 1..n {
        let t = fs[i].1;
        ans = if fs[i].0 == fs[0].0 {
            ans.max(s + t / 2)
        } else {
            ans.max(s + t)
        }
    }

    println!("{}", ans);
}
